pkg/fanal/analyzer/language/nodejs/yarn/yarn.go

```diff
@@ -242,3 +275,157 @@ func (a yarnAnalyzer) traverseWorkspaces(fsys fs.FS, workspaces []string) ([]pac
 	}
 	return pkgs, nil
 }
+
+type traverseFunc func(pkg packagejson.Package) error
+
+func (a yarnAnalyzer) traversePkgs(fsys fs.FS, lockPath string, fn traverseFunc) error {
+	dir := filepath.Dir(lockPath)
+
+	nodeModulesPath := path.Join(dir, "node_modules")
+
+	if _, err := fs.Stat(fsys, nodeModulesPath); errors.Is(err, fs.ErrNotExist) {
+		// try to find for yarn v2+
+		return a.traverseYarnModernPkgs(fsys, dir, fn)
+	} else if err != nil {
+		return xerrors.Errorf("unable to parse %q: %w", nodeModulesPath, err)
+	}
+
+	return a.traverseYarnClassicPkgs(fsys, nodeModulesPath, fn)
+}
+
+func isNodeModulesPkg(path string, _ fs.DirEntry) bool {
+	// We can't check for sure whether the package.json is valid in `Required` fn.
+	// For example, package.json can be located near yarn.lock, then it fits,
+	// or in the package tests in node_modules - then not.
+	// That's why we add all the package.json and filter them later.
+	dirs, fileName := splitPath(path)
+	nodeModulesIdx := len(dirs) - 2
+	// the scope starts with "@" https://docs.npmjs.com/cli/v9/using-npm/scope
+	if len(dirs) > 2 && strings.HasPrefix(dirs[len(dirs)-2], "@") {
+		nodeModulesIdx -= 1
+	}
+	// The file path to package.json - */node_modules/<package_name>/package.json
+	// or */node_modules/@<scope_name>/<package_name>/package.json
+	return len(dirs) > 1 && dirs[nodeModulesIdx] == "node_modules" && fileName == types.NpmPkg
+}
```

Comment:
Do we need this complex logic?

>or in the package tests in node_modules - then not.

Do you mean dummy package.json?

My question is why we cannot simply check if the file name matches `package.json`.

Reply:
@knqyf263 It can be a dummy file for tests.

Reply:
Yes, I agree